### PR TITLE
Zilinghan/refactor

### PR DESCRIPTION
**Pull request by Zilinghan**
refactoring APPFL

### CODE REVIEW

**Comment by Zilinghan** [comment]
Update 1 - Refactor the `examples/dataloader` folder. Instead of `iid.py`, `class_noiid.py` and `dirichlet_noiid.py`, now we have dataloader for each dataset: `mnist_dataloader.py`, `cifar_dataloader.py`, `flamby_dataloader.py`, and `nrel_dataloder.py`, where `mnist_dataloader.py` and `cifar_dataloader.py` take an argument called `partition` to specify the partition strategy.

Update 2 - Rename all `gc` to `globus_compute`

**Comment by yim0331** [comment]
Update: 
1. Add femnist_dataloader.py and change femnist_mpi.py and femnist_grpc.py to call the dataloader
2. Add celeba_dataloader.py and change celeba.py to call the dataloader
3. Add loss_fn for femnist_grpc.py

@Zilinghan I'm having "FLamby is not installed" message to run the femnist_mpi.py and celeba.py. The message is printed from examples/models/util.py while importing flamby dependencies. I think it's not necessary to show this message.

**Comment by Zilinghan** [comment]
Update 3 - Change the base class definition for general FL algorithms (`FedAvg`, `FedAvgM` for server, and `ClientOptim`, `ClientStepOptim` for client for example) to remove the unnecessary class attributes such as penalty, dual, etc.
Create new base classes `PPFLServer` and `PPFLClient` for privacy-preserving FL algorithms. This change is mainly used for increasing the readability of the codes for users.

**Comment by Zilinghan** [comment]
Update 4 - Add privacy preserving parameters in the examples, namely, 

- `use_dp`: whether to use differential privacy
- `epsilon`: privacy budget
- `clip_grad`: whether to clip the gradients
- `clip_value`: max norm of the gradient
- `clip_norm`: type of the used p-norm for gradient clipping

**Comment by Zilinghan** [comment]
> Update:
> 
> 1. Add femnist_dataloader.py and change femnist_mpi.py and femnist_grpc.py to call the dataloader
> 2. Add celeba_dataloader.py and change celeba.py to call the dataloader
> 3. Add loss_fn for femnist_grpc.py
> 
> @Zilinghan I'm having "FLamby is not installed" message to run the femnist_mpi.py and celeba.py. The message is printed from examples/models/util.py while importing flamby dependencies. I think it's not necessary to show this message.

Resolved "Flamby is not installed" message
